File: Sources/ZarliAdapterAdMob/ZarliAdMobMediationInterstitialAd.m
#import "ZarliAdMobMediationInterstitialAd.h"
@import ZarliSDKSwift;

@interface ZarliAdMobMediationInterstitialAd () <ZarliInterstitialAdDelegate>
@end

@implementation ZarliAdMobMediationInterstitialAd {
  GADMediationInterstitialAdConfiguration *_adConfiguration;
  GADMediationInterstitialLoadCompletionHandler _completionHandler;
  __weak id<GADMediationInterstitialAdEventDelegate> _delegate;
  ZarliInterstitialAd *_zarliAd;
}

- (instancetype)initWithAdConfiguration:
                    (GADMediationInterstitialAdConfiguration *)adConfiguration
                      completionHandler:
                          (GADMediationInterstitialLoadCompletionHandler)
                              completionHandler {
  self = [super init];
  if (self) {
    _adConfiguration = adConfiguration;
    _completionHandler = completionHandler;
  }
  return self;
}

- (void)loadAd {
  NSString *adUnitId = @"default-interstitial";
  double bidFloor = 0.0;

  // Parse ad unit configuration from AdMob
  id parameterObj = _adConfiguration.credentials.settings[@"parameter"];
  if (parameterObj && [parameterObj isKindOfClass:[NSString class]]) {
    NSString *parameter = (NSString *)parameterObj;
    NSData *data = [parameter dataUsingEncoding:NSUTF8StringEncoding];
    if (data) {
      NSError *error = nil;
      NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data
                                                           options:0
                                                             error:&error];
      if (json && [json isKindOfClass:[NSDictionary class]]) {
        if (json[@"adUnitId"]) {
          adUnitId = json[@"adUnitId"];
        }
        if (json[@"bidFloor"]) {
          bidFloor = [json[@"bidFloor"] doubleValue];
        }
      } else {
        // Fallback: treat entire parameter as ad unit ID
        adUnitId = parameter;
      }
    }
  }

  // Initialize Zarli Interstitial Ad
  _zarliAd = [[ZarliInterstitialAd alloc] initWithAdUnitId:adUnitId];
  _zarliAd.bidFloor = bidFloor;
  _zarliAd.delegate = self;
  [_zarliAd load];
}

#pragma mark - GADMediationInterstitialAd Protocol

- (void)presentFromViewController:(UIViewController *)viewController {
  if (_zarliAd.isReady) {
    [_zarliAd showFrom:viewController];
  } else {
    NSError *error = [NSError
        errorWithDomain:@"com.zarli.adapter"
                   code:0
               userInfo:@{NSLocalizedDescriptionKey : @"Ad not ready"}];
    if ([_delegate respondsToSelector:@selector(didFailToPresentWithError:)]) {
      [_delegate didFailToPresentWithError:error];
    }
  }
}

#pragma mark - ZarliInterstitialAdDelegate

- (void)adDidLoad:(ZarliInterstitialAd *)ad {
  if (_completionHandler) {
    _delegate = _completionHandler(self, nil);
    _completionHandler = nil;
  }
}

- (void)ad:(ZarliInterstitialAd *)ad didFailToLoad:(NSError *)error {
  if (_completionHandler) {
    _completionHandler(nil, error);
    _completionHandler = nil;
  }
}

- (void)adDidShow:(ZarliInterstitialAd *)ad {
  if ([_delegate respondsToSelector:@selector(willPresentFullScreenView)]) {
    [_delegate willPresentFullScreenView];
  }
  if ([_delegate respondsToSelector:@selector(reportImpression)]) {
    [_delegate reportImpression];
  }
}

- (void)adDidDismiss:(ZarliInterstitialAd *)ad {
  if ([_delegate respondsToSelector:@selector(willDismissFullScreenView)]) {
    [_delegate willDismissFullScreenView];
  }
  if ([_delegate respondsToSelector:@selector(didDismissFullScreenView)]) {
    [_delegate didDismissFullScreenView];
  }
}

- (void)adDidClick:(ZarliInterstitialAd *)ad {
  if ([_delegate respondsToSelector:@selector(reportClick)]) {
    [_delegate reportClick];
  }
}

@end

File: Sources/ZarliAdapterAdMob/ZarliAdMobMediationInterstitialAd.swift
import Foundation
import GoogleMobileAds
import ZarliSDKSwift

public class ZarliAdMobMediationInterstitialAd: NSObject, GoogleMobileAds.MediationInterstitialAd {
    
    private let adConfiguration: GoogleMobileAds.MediationInterstitialAdConfiguration
    private var completionHandler: GoogleMobileAds.MediationInterstitialLoadCompletionHandler?
    private weak var delegate: GoogleMobileAds.MediationInterstitialAdEventDelegate?
    private var zarliAd: ZarliInterstitialAd?
    
    public init(configuration: GoogleMobileAds.MediationInterstitialAdConfiguration, completionHandler: @escaping GoogleMobileAds.MediationInterstitialLoadCompletionHandler) {
        self.adConfiguration = configuration
        self.completionHandler = completionHandler
    }
    
    public func loadAd() {
        print("ZarliAdapter: Interstitial loadAd() called")
        var adUnitId = "default-interstitial"
        var explicitBidFloor: Double?
        
        // 1. Parse 'parameter' string (supports JSON or raw string)
        if let parameter = adConfiguration.credentials.settings["parameter"] as? String {
             print("ZarliAdapter: Received parameter: \(parameter)")
            
            var json: [String: Any]?
            if let data = parameter.data(using: .utf8) {
                json = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
            }
            
            // Fallback: If not valid JSON but contains colons, try wrapping in braces
            if json == nil && parameter.contains(":") {
                let wrapped = "{" + parameter + "}"
                if let data = wrapped.data(using: .utf8) {
                    json = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
                }
            }

            if let finalJson = json {
                // Handle JSON format
                if let unitId = finalJson["adUnitId"] as? String {
                    adUnitId = unitId
                }
                explicitBidFloor = finalJson["bidFloor"] as? Double
            } else {
                // Fallback: Treat entire string as Ad Unit ID
                adUnitId = parameter
            }
        } else {
             print("ZarliAdapter: No parameter found")
        }
        
        // 2. Determine Bid Floor
        let finalBidFloor: Double
        if let floor = explicitBidFloor {
            finalBidFloor = floor
        } else {
            finalBidFloor = 0.0
        }
        
        print("ZarliAdapter: Using AdUnitID: \(adUnitId), BidFloor: \(finalBidFloor)")
        
        zarliAd = ZarliInterstitialAd(adUnitId: adUnitId)
        zarliAd?.bidFloor = finalBidFloor
        zarliAd?.delegate = self
        zarliAd?.load()
    }
    
    public func present(from viewController: UIViewController) {
        if let ad = zarliAd, ad.isReady {
            ad.show(from: viewController)
        } else {
            // Should not happen if lifecycle is correct
            // But if it does, let the delegate know?
            let error = NSError(domain: "com.zarli.sdk", code: 0, userInfo: [NSLocalizedDescriptionKey: "Ad not ready"])
            delegate?.didFailToPresentWithError(error)
        }
    }
}

extension ZarliAdMobMediationInterstitialAd: ZarliInterstitialAdDelegate {
    public func adDidLoad(_ ad: ZarliInterstitialAd) {
        delegate = completionHandler?(self, nil)
    }
    
    public func ad(_ ad: ZarliInterstitialAd, didFailToLoad error: Error) {
        _ = completionHandler?(nil, error)
    }
    
    public func adDidShow(_ ad: ZarliInterstitialAd) {
        delegate?.willPresentFullScreenView()
        delegate?.reportImpression()
    }
    
    public func adDidDismiss(_ ad: ZarliInterstitialAd) {
        delegate?.willDismissFullScreenView()
        delegate?.didDismissFullScreenView()
    }
    
    public func adDidClick(_ ad: ZarliInterstitialAd) {
        delegate?.reportClick()
    }
}
